docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+	Since this number also includes the container’s writable layer, we will have to somehow subtract that usage from the overall usage of the storage driver directory. Luckily, we can use the same mechanism for tracking container’s writable layer. Once we apply a different `gid` to the container’s writable layer, which is located under `/var/lib/docker/<storage_driver>/diff/<container_id>`, the quota subsystem will not include the container’s writable layer usage.
+
+**Pros**:
+
+* Low overhead tracking provided by the kernel.
+
+
+**Cons**
+
+* Requires updates to default ownership on docker’s internal storage driver directories. We will have to deal with storage driver implementation implementation details in any approach that is not docker native.
+
+* Requires additional node configuration - quota subsystem needs to be setup on the node. This can either be automated or made a requirement for the node.
+
+* Kubelet needs to perform gid management. A range of gids have to allocated to the kubelet for the purposes of quota management. This range must not be used for any other purposes out of band.
+
+* Breaks `docker save` semantics. Since kubernetes assumes immutable images, this is not a blocker. To support quota in docker, we will need user-namespaces along with custom gid mapping for each container. This feature does not exist today.
+
+*Note: Refer to the [Appendix](#appendix) section more real examples on using quota with docker.*
+
+
+3. Enhance docker
+
+	Docker handles the upload and download of image layers. It can embed enough information about each layer. If docker is enhanced to expose this information, we can statically identify space about to be occupied by read-only image layers, even before the image layers are downloaded.
+
+	A new [docker feature](https://github.com/docker/docker/pull/16450) (docker pull --dry-run) is pending review, which outputs the disk space that will be consumed by new images. Once this feature lands, we can perform feasibility checks and reject pods that will consume more disk space that what is current availability on the node.
+
+	We can also attempt to use this feature to determine image layer’s disk usage statically. The metadata files that docker creates for these layers will not be tracked though.
+
+**Pros**
+
+* Works across all storage drivers since docker abstracts the storage drivers.
+
+* Less code to maintain in kubelet.
+
+**Cons**
+
+* Not available today.
+
+* Requires serialized image pulls.
+
+* Metadata files are not tracked.
+
+
+##### Devicemapper
+
+Devicemapper storage driver will setup two volumes, metadata and data, that will be used to store image layers and container writable layer. The volumes can be real devices or loopback. A Pool device is created which uses the underlying volume for real storage.
+
```

Comment:
We strongly discourage usage of loopback devices for thin pool creation. User should be setting up an lvm thin pool from real block devices and pass that pool to docker. I want to make sure that we don't get too hung up on loopback or promote it as supported model.

Following is link to docker-storage-setup utility which an setup an lvm thin pool relatively easily as long as one can point this utility to a volume group with free space or pass in a disk to use and utility will create a volume group and create an lvm thin pool out of it.

https://github.com/projectatomic/docker-storage-setup